

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>CITEgeist.model.utils &#8212; CITEgeist  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/CITEgeist/model/utils';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    <p class="title logo__title">CITEgeist  documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API Reference</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/leeoesterreich/CITEgeist" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <h1>Source code for CITEgeist.model.utils</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">gc</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">jensenshannon</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">pearsonr</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">mean_squared_error</span><span class="p">,</span> <span class="n">mean_absolute_error</span>

<div class="viewcode-block" id="validate_cell_profile_dict"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.validate_cell_profile_dict">[docs]</a><span class="k">def</span> <span class="nf">validate_cell_profile_dict</span><span class="p">(</span><span class="n">cell_profile_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Validate the structure of a cell profile dictionary.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cell_profile_dict</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">return</span> <span class="nb">all</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">cell_profile_dict</span><span class="o">.</span><span class="n">items</span><span class="p">())</span></div>

<div class="viewcode-block" id="save_results_to_output"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.save_results_to_output">[docs]</a><span class="k">def</span> <span class="nf">save_results_to_output</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">filepath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save results as a CSV file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span></div>

<div class="viewcode-block" id="cleanup_memory"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.cleanup_memory">[docs]</a><span class="k">def</span> <span class="nf">cleanup_memory</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Force garbage collection to free memory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gc</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span></div>

<div class="viewcode-block" id="setup_logging"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.setup_logging">[docs]</a><span class="k">def</span> <span class="nf">setup_logging</span><span class="p">(</span><span class="n">output_folder</span><span class="p">,</span> <span class="n">sample_name</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set up dynamic logging.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">log_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_folder</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_name</span><span class="si">}</span><span class="s2">_CITEgeist.log&quot;</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
        <span class="n">filename</span><span class="o">=</span><span class="n">log_file</span><span class="p">,</span>
        <span class="n">filemode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">,</span>
        <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> - </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">,</span>
        <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span>
    <span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Logging initialized.&#39;</span><span class="p">)</span></div>

<span class="c1">### üìè **Spatial Neighbor Functions**</span>



<div class="viewcode-block" id="find_fixed_radius_neighbors"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.find_fixed_radius_neighbors">[docs]</a><span class="k">def</span> <span class="nf">find_fixed_radius_neighbors</span><span class="p">(</span><span class="n">spot_index</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">50</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find neighbors within a fixed radius of a given spot.</span>

<span class="sd">    Args:</span>
<span class="sd">        spot_index (int): Index of the central spot in the AnnData object.</span>
<span class="sd">        adata (AnnData): Spatial transcriptomics dataset with obsm[&#39;spatial&#39;] containing spot coordinates.</span>
<span class="sd">        radius (float): Fixed radius to identify neighboring spots.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: (central_spot_name, list of neighbor spot names)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">coordinates</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span>
    <span class="n">central_coord</span> <span class="o">=</span> <span class="n">coordinates</span><span class="p">[</span><span class="n">spot_index</span><span class="p">]</span>
    
    <span class="c1"># Identify all spots within the given radius, excluding the central spot</span>
    <span class="n">neighbors</span> <span class="o">=</span> <span class="p">[</span><span class="n">idx</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">coord</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">coordinates</span><span class="p">)</span>
                 <span class="k">if</span> <span class="n">idx</span> <span class="o">!=</span> <span class="n">spot_index</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">coord</span> <span class="o">-</span> <span class="n">central_coord</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">radius</span><span class="p">]</span>
    
    <span class="c1"># Convert indices to spot names</span>
    <span class="n">neighbor_names</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">neighbors</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">central_spot_name</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">spot_index</span><span class="p">]</span>
    
    <span class="k">return</span> <span class="n">central_spot_name</span><span class="p">,</span> <span class="n">neighbor_names</span><span class="p">,</span> <span class="n">spot_index</span><span class="p">,</span> <span class="n">neighbors</span></div>


<div class="viewcode-block" id="get_neighbors_with_fixed_radius"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.get_neighbors_with_fixed_radius">[docs]</a><span class="k">def</span> <span class="nf">get_neighbors_with_fixed_radius</span><span class="p">(</span><span class="n">spot_index</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">include_center</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span> <span class="c1"># In Visium, 2 rings gives you the adjacent 6 units</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get indices of neighboring spots based on a fixed radius around the central spot.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - spot_index (int): The index of the central spot.</span>
<span class="sd">    - adata (AnnData): The AnnData object with spatial coordinates in obsm[&#39;spatial&#39;].</span>
<span class="sd">    - radius (float): Fixed radius for finding neighbors.</span>
<span class="sd">    - include_center (bool): Whether to include the central spot in the neighbor list.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - List of indices representing the central spot and its neighbors.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Find neighbors within the given radius</span>
    <span class="n">central_spot_names</span><span class="p">,</span> <span class="n">neighbor_spots_names</span><span class="p">,</span> <span class="n">spot_index</span><span class="p">,</span> <span class="n">neighbors</span> <span class="o">=</span> <span class="n">find_fixed_radius_neighbors</span><span class="p">(</span><span class="n">spot_index</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="p">)</span>
    
    <span class="c1"># Optionally include the central spot itself</span>
    <span class="k">if</span> <span class="n">include_center</span><span class="p">:</span>
        <span class="n">neighbors</span> <span class="o">=</span> <span class="p">[</span><span class="n">spot_index</span><span class="p">]</span> <span class="o">+</span> <span class="n">neighbors</span>
    
    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total neighbors for spot </span><span class="si">{</span><span class="n">spot_index</span><span class="si">}</span><span class="s2"> within radius </span><span class="si">{</span><span class="n">radius</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">neighbors</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">neighbors</span></div>
<div class="viewcode-block" id="plot_neighbors_with_fixed_radius"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.plot_neighbors_with_fixed_radius">[docs]</a><span class="k">def</span> <span class="nf">plot_neighbors_with_fixed_radius</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">num_spots</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot neighbors for multiple random central spots using `sc.pl.spatial`.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): Spatial transcriptomics dataset with obsm[&#39;spatial&#39;].</span>
<span class="sd">        radius (float): Fixed radius to identify neighboring spots.</span>
<span class="sd">        num_spots (int): Number of random spots to visualize.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: Displays a series of spatial plots showing neighbors.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">random</span>
    
    <span class="c1"># Select random spots</span>
    <span class="n">random_spots</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">min</span><span class="p">(</span><span class="n">num_spots</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    
    <span class="c1"># Define colorblind-friendly colors that contrast well with orange background</span>
    <span class="c1"># Using dark blue, white, and black for maximum contrast</span>
    <span class="n">color_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Other spots&#39;</span><span class="p">:</span> <span class="s1">&#39;#00FF00&#39;</span><span class="p">,</span>  <span class="c1"># Green</span>
        <span class="s1">&#39;Neighbor&#39;</span><span class="p">:</span> <span class="s1">&#39;#40E0D0&#39;</span><span class="p">,</span>     <span class="c1"># Turquoise</span>
        <span class="s1">&#39;Central spot&#39;</span><span class="p">:</span> <span class="s1">&#39;#0000FF&#39;</span>  <span class="c1"># Dark blue</span>
    <span class="p">}</span>
    
    <span class="k">for</span> <span class="n">spot_index</span> <span class="ow">in</span> <span class="n">random_spots</span><span class="p">:</span>
        <span class="c1"># Find neighbors within the given radius</span>
        <span class="n">central_spot_names</span><span class="p">,</span> <span class="n">neighbor_spots_names</span><span class="p">,</span> <span class="n">spot_index</span><span class="p">,</span> <span class="n">neighbors</span> <span class="o">=</span> <span class="n">find_fixed_radius_neighbors</span><span class="p">(</span><span class="n">spot_index</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="p">)</span>

        <span class="c1"># Create a temporary column to highlight spots</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;highlight&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Other spots&#39;</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">neighbor_spots_names</span><span class="p">,</span> <span class="s1">&#39;highlight&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Neighbor&#39;</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">central_spot_names</span><span class="p">,</span> <span class="s1">&#39;highlight&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Central spot&#39;</span>
        
        <span class="c1"># Plot using `sc.pl.spatial` with custom colors</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;highlight&#39;</span><span class="p">,</span>
            <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Neighbors within </span><span class="si">{</span><span class="n">radius</span><span class="si">}</span><span class="s2"> units for Spot </span><span class="si">{</span><span class="n">central_spot_names</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">spot_size</span> <span class="o">=</span> <span class="mi">75</span><span class="p">,</span>
            <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">palette</span><span class="o">=</span><span class="n">color_dict</span>
        <span class="p">)</span>
        
        <span class="c1"># Clean up temporary column after each plot</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;highlight&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>
        
<div class="viewcode-block" id="assert_neighborhood_size"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.assert_neighborhood_size">[docs]</a><span class="k">def</span> <span class="nf">assert_neighborhood_size</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_profile_dict</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">num_spots</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">random</span>
    
    <span class="c1"># Select random spots</span>
    <span class="n">random_spots</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">min</span><span class="p">(</span><span class="n">num_spots</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    
    <span class="n">neighborhood_sizes</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">spot_index</span> <span class="ow">in</span> <span class="n">random_spots</span><span class="p">:</span>
        <span class="c1"># Find neighbors within the given radius</span>
        <span class="n">central_spot_names</span><span class="p">,</span> <span class="n">neighbor_spots_names</span><span class="p">,</span> <span class="n">spot_index</span><span class="p">,</span> <span class="n">neighbors</span> <span class="o">=</span> <span class="n">find_fixed_radius_neighbors</span><span class="p">(</span><span class="n">spot_index</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="p">)</span>

    <span class="n">central_spot_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">central_spot_names</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">central_spot_names</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="k">else</span> <span class="n">central_spot_names</span>
    <span class="n">neighbor_spots_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">neighbor_spots_names</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">neighbor_spots_names</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="k">else</span> <span class="n">neighbor_spots_names</span>

    <span class="n">neighborhood_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">central_spot_names</span> <span class="o">+</span> <span class="n">neighbor_spots_names</span><span class="p">)</span>
    
    
    <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_profile_dict</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">neighborhood_sizes</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;Some neighborhood values in the list are less than </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">cell_profile_dict</span><span class="p">)</span><span class="si">}</span><span class="s2"> celltypes being deconvoluted&quot;</span></div>

<div class="viewcode-block" id="benchmark_cell_proportions"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.benchmark_cell_proportions">[docs]</a><span class="k">def</span> <span class="nf">benchmark_cell_proportions</span><span class="p">(</span><span class="n">true_proportions</span><span class="p">,</span> <span class="n">predicted_proportions</span><span class="p">,</span> <span class="n">cell_type_names</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate performance metrics for cell type proportion predictions.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        true_proportions (np.ndarray): Ground truth cell type proportions matrix</span>
<span class="sd">        predicted_proportions (np.ndarray): Predicted cell type proportions matrix</span>
<span class="sd">        cell_type_names (list): Names of cell types corresponding to matrix columns</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary containing various performance metrics</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">true_proportions</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">predicted_proportions</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input proportions must be numpy arrays&quot;</span><span class="p">)</span>

    <span class="c1"># Initialize JSD matrix</span>
    <span class="n">true_jsd_mtrx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">true_proportions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span>

    <span class="c1"># Calculate Jensen-Shannon Divergence</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">true_proportions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">true_proportions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:],</span> <span class="n">predicted_proportions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]])</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">predicted_proportions</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">true_jsd_mtrx</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">jensenshannon</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">base</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">true_jsd_mtrx</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="c1"># Calculate per-celltype and overall metrics</span>
    <span class="n">RMSE</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">MAE</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">all_rmse</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">all_mae</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">true_proportions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">mse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">true_proportions</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">predicted_proportions</span><span class="p">[:,</span> <span class="n">i</span><span class="p">])</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">all_rmse</span> <span class="o">+=</span> <span class="n">mse</span>
        <span class="n">RMSE</span><span class="p">[</span><span class="n">cell_type_names</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mse</span> <span class="o">/</span> <span class="n">true_proportions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="n">mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">true_proportions</span><span class="p">[:,</span> <span class="n">i</span><span class="p">],</span> <span class="n">predicted_proportions</span><span class="p">[:,</span> <span class="n">i</span><span class="p">])</span>
        <span class="n">all_mae</span> <span class="o">+=</span> <span class="n">mae</span>
        <span class="n">MAE</span><span class="p">[</span><span class="n">cell_type_names</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">=</span> <span class="n">mae</span>

    <span class="c1"># Calculate overall metrics</span>
    <span class="n">all_rmse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">all_rmse</span> <span class="o">/</span> <span class="p">(</span><span class="n">true_proportions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">true_proportions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">all_mae</span> <span class="o">=</span> <span class="n">all_mae</span> <span class="o">/</span> <span class="n">true_proportions</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Calculate JSD quantiles and correlation</span>
    <span class="n">quants_jsd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">true_jsd_mtrx</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">])</span>
    <span class="n">corr</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span><span class="n">true_proportions</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">predicted_proportions</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s1">&#39;JSD&#39;</span><span class="p">:</span> <span class="n">quants_jsd</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
        <span class="s1">&#39;RMSE&#39;</span><span class="p">:</span> <span class="n">RMSE</span><span class="p">,</span>
        <span class="s1">&#39;Sum_RMSE&#39;</span><span class="p">:</span> <span class="n">all_rmse</span><span class="p">,</span>
        <span class="s1">&#39;MAE&#39;</span><span class="p">:</span> <span class="n">MAE</span><span class="p">,</span>
        <span class="s1">&#39;Sum_MAE&#39;</span><span class="p">:</span> <span class="n">all_mae</span><span class="p">,</span>
        <span class="s1">&#39;corr&#39;</span><span class="p">:</span> <span class="n">corr</span>
    <span class="p">}</span></div>

<div class="viewcode-block" id="export_anndata_layers"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.export_anndata_layers">[docs]</a><span class="k">def</span> <span class="nf">export_anndata_layers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">pass_number</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Export all layers of an AnnData object to separate CSV files.</span>
<span class="sd">    Creates separate folders for different passes.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): AnnData object containing the layers to export</span>
<span class="sd">        output_dir (str): Base directory where CSV files will be saved</span>
<span class="sd">        pass_number (int, optional): If specified, only export layers from this pass</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create base output directory</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Create pass-specific directory if needed</span>
    <span class="k">if</span> <span class="n">pass_number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">target_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pass</span><span class="si">{</span><span class="n">pass_number</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">target_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">target_dir</span> <span class="o">=</span> <span class="n">output_dir</span>

    <span class="c1"># Filter layers for specific pass if requested</span>
    <span class="n">layer_pattern</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;_pass</span><span class="si">{</span><span class="n">pass_number</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">pass_number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
    
    <span class="k">for</span> <span class="n">layer_name</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="c1"># Skip if not matching pass number</span>
        <span class="k">if</span> <span class="n">layer_pattern</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">layer_pattern</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">layer_name</span><span class="p">:</span>
            <span class="k">continue</span>
        
        <span class="c1"># Extract data and ensure it&#39;s dense</span>
        <span class="n">layer_data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">layer_name</span><span class="p">]</span>
        <span class="n">dense_data</span> <span class="o">=</span> <span class="n">layer_data</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">layer_data</span><span class="p">,</span> <span class="s1">&#39;toarray&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">layer_data</span>
        
        <span class="c1"># Create DataFrame</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dense_data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        
        <span class="c1"># Extract cell type name from layer name consistently</span>
        <span class="n">cell_type</span> <span class="o">=</span> <span class="n">layer_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_genes_pass&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        
        <span class="c1"># Save with standardized naming including pass number</span>
        <span class="k">if</span> <span class="n">pass_number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">_layer_pass</span><span class="si">{</span><span class="n">pass_number</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">_layer.csv&quot;</span><span class="p">)</span>
            
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Exported layer &#39;</span><span class="si">{</span><span class="n">layer_name</span><span class="si">}</span><span class="s2">&#39; to &#39;</span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="calculate_expression_metrics"><a class="viewcode-back" href="../../../api.html#CITEgeist.model.utils.calculate_expression_metrics">[docs]</a><span class="k">def</span> <span class="nf">calculate_expression_metrics</span><span class="p">(</span><span class="n">ground_truth_dir</span><span class="p">,</span> <span class="n">predictions_dir</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="s1">&#39;range&#39;</span><span class="p">,</span> <span class="n">pass_number</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate performance metrics for gene expression predictions.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        ground_truth_dir (str): Directory containing ground truth CSV files</span>
<span class="sd">        predictions_dir (str): Directory containing prediction CSV files</span>
<span class="sd">        normalize (str): Normalization method for NRMSE (&#39;range&#39; or &#39;mean&#39;)</span>
<span class="sd">        pass_number (int, optional): If specified, look for predictions in pass-specific subdirectory</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary containing performance metrics per cell type and overall statistics</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">metrics_per_cell_type</span> <span class="o">=</span> <span class="p">{}</span>
    
    <span class="c1"># Adjust predictions directory if pass number specified</span>
    <span class="k">if</span> <span class="n">pass_number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">predictions_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">predictions_dir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;pass</span><span class="si">{</span><span class="n">pass_number</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ground truth directory: </span><span class="si">{</span><span class="n">ground_truth_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ground truth files: </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">ground_truth_dir</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Layer directory: </span><span class="si">{</span><span class="n">predictions_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Layer files: </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">predictions_dir</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Get sorted lists of files with pass number handling</span>
    <span class="n">gt_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">ground_truth_dir</span><span class="p">)</span> <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;_GT.csv&#39;</span><span class="p">)])</span>
    <span class="k">if</span> <span class="n">pass_number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pred_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">predictions_dir</span><span class="p">)</span> <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;_layer_pass</span><span class="si">{</span><span class="n">pass_number</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">)])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pred_files</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">predictions_dir</span><span class="p">)</span> <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;_layer.csv&#39;</span><span class="p">)])</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;GT files: &quot;</span><span class="p">,</span> <span class="n">gt_files</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Pred files: &quot;</span><span class="p">,</span> <span class="n">pred_files</span><span class="p">)</span>

    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">gt_files</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">pred_files</span><span class="p">),</span> <span class="s2">&quot;Number of ground truth files and prediction files do not match&quot;</span>

    <span class="c1"># Create a dictionary to map cell types to their ground truth files</span>
    <span class="n">gt_file_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_GT.csv&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">):</span> <span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">gt_files</span><span class="p">}</span>

    <span class="c1"># Create a list to store matched prediction and ground truth file pairs</span>
    <span class="n">matched_files</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">pred_file</span> <span class="ow">in</span> <span class="n">pred_files</span><span class="p">:</span>
        <span class="c1"># Remove pass number suffix if present</span>
        <span class="n">base_pred_file</span> <span class="o">=</span> <span class="n">pred_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;_pass</span><span class="si">{</span><span class="n">pass_number</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">pass_number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">pred_file</span>
        <span class="n">cell_type</span> <span class="o">=</span> <span class="n">base_pred_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_layer.csv&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">gt_file_map</span><span class="p">:</span>
            <span class="n">matched_files</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">pred_file</span><span class="p">,</span> <span class="n">gt_file_map</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]))</span>

    <span class="c1"># Sort matched files by the cell type name</span>
    <span class="n">matched_files</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># Calculate metrics</span>
    <span class="k">for</span> <span class="n">pred_filename</span><span class="p">,</span> <span class="n">gt_filename</span> <span class="ow">in</span> <span class="n">matched_files</span><span class="p">:</span>
        <span class="n">cell_type</span> <span class="o">=</span> <span class="n">gt_filename</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_GT.csv&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">gt_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ground_truth_dir</span><span class="p">,</span> <span class="n">gt_filename</span><span class="p">)</span>
        <span class="n">pred_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">predictions_dir</span><span class="p">,</span> <span class="n">pred_filename</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">pred_filepath</span><span class="p">):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Prediction file for </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2"> not found. Skipping.&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># Load and preprocess data</span>
        <span class="n">gt_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gt_filepath</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">pred_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pred_filepath</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Find common genes and spots</span>
        <span class="n">common_genes</span> <span class="o">=</span> <span class="n">gt_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">pred_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="n">common_spots</span> <span class="o">=</span> <span class="n">gt_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">pred_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">common_genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">common_spots</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No common genes or spots for </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">. Skipping.&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1"># Subset and normalize data</span>
        <span class="n">gt_subset</span> <span class="o">=</span> <span class="n">gt_df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">common_genes</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">common_spots</span><span class="p">)</span>
        <span class="n">pred_subset</span> <span class="o">=</span> <span class="n">pred_df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">common_genes</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">common_spots</span><span class="p">)</span>
        
        <span class="n">gt_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">gt_subset</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">common_genes</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">common_spots</span><span class="p">)</span>
        <span class="n">pred_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">pred_subset</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">common_genes</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">common_spots</span><span class="p">)</span>

        <span class="c1"># Calculate metrics</span>
        <span class="n">mse</span> <span class="o">=</span> <span class="n">mean_squared_error</span><span class="p">(</span><span class="n">gt_df</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">pred_df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
        <span class="n">rmse</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mse</span><span class="p">)</span>
        <span class="n">mae</span> <span class="o">=</span> <span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">gt_df</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">pred_df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

        <span class="c1"># Calculate NRMSE</span>
        <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="s1">&#39;range&#39;</span><span class="p">:</span>
            <span class="n">range_gt</span> <span class="o">=</span> <span class="n">gt_df</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">gt_df</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
            <span class="n">nrmse</span> <span class="o">=</span> <span class="n">rmse</span> <span class="o">/</span> <span class="n">range_gt</span> <span class="k">if</span> <span class="n">range_gt</span> <span class="o">!=</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="k">elif</span> <span class="n">normalize</span> <span class="o">==</span> <span class="s1">&#39;mean&#39;</span><span class="p">:</span>
            <span class="n">mean_gt</span> <span class="o">=</span> <span class="n">gt_df</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
            <span class="n">nrmse</span> <span class="o">=</span> <span class="n">rmse</span> <span class="o">/</span> <span class="n">mean_gt</span> <span class="k">if</span> <span class="n">mean_gt</span> <span class="o">!=</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Normalization type must be &#39;range&#39; or &#39;mean&#39;&quot;</span><span class="p">)</span>

        <span class="c1"># Assertions that also print the celltype</span>
        <span class="k">assert</span> <span class="n">nrmse</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">nrmse</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;NRMSE is None for </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">assert</span> <span class="n">rmse</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rmse</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;RMSE is None for </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">assert</span> <span class="n">mae</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">mae</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;MAE is None for </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">&quot;</span>

        <span class="n">metrics_per_cell_type</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;RMSE&#39;</span><span class="p">:</span> <span class="n">rmse</span><span class="p">,</span> <span class="s1">&#39;NRMSE&#39;</span><span class="p">:</span> <span class="n">nrmse</span><span class="p">,</span> <span class="s1">&#39;MAE&#39;</span><span class="p">:</span> <span class="n">mae</span><span class="p">}</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Metrics for </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">: RMSE=</span><span class="si">{</span><span class="n">rmse</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">, NRMSE=</span><span class="si">{</span><span class="n">nrmse</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">, MAE=</span><span class="si">{</span><span class="n">mae</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
    <span class="k">return</span> <span class="n">metrics_per_cell_type</span></div>


</pre></div>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lee/Oesterreich Lab
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      ¬© Copyright 2025, Lee/Oesterreich Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>