

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>API Reference &#8212; CITEgeist  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributing" href="contributing.html" />
    <link rel="prev" title="Vignette 4: Integration with External Tools" href="notebooks/vignette_4_external_tools.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  <div class="navbar-header-items__start">
    
      <div class="navbar-item">
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    <p class="title logo__title">CITEgeist  documentation</p>
  
</a></div>
    
  </div>
  
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="installation.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="quickstart.html">
                        Quick Start
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="tutorial.html">
                        Tutorial
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="examples.html">
                        Examples
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="notebooks.html">
                        Interactive Tutorials and Vignettes
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        API Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="changelog.html">
                        Changelog
                      </a>
                    </li>
                
                </div>
            </div>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
        </div>
      
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/leeoesterreich/CITEgeist" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">
<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item"><nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="installation.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="quickstart.html">
                        Quick Start
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="tutorial.html">
                        Tutorial
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="examples.html">
                        Examples
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="notebooks.html">
                        Interactive Tutorials and Vignettes
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        API Reference
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="changelog.html">
                        Changelog
                      </a>
                    </li>
                
                </div>
            </div>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/leeoesterreich/CITEgeist" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumbs">
  <ul class="bd-breadcrumbs" role="navigation" aria-label="Breadcrumb">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page">API Reference</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="api-reference">
<h1>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this heading">#</a></h1>
<p>This page provides an auto-generated summary of CITEgeist’s public API.</p>
<section id="citegeist-model">
<h2>CITEgeist Model<a class="headerlink" href="#citegeist-model" title="Permalink to this heading">#</a></h2>
<p>The main CITEgeist model class for spatial transcriptomics analysis.</p>
<span class="target" id="module-citegeist_model"></span><dl class="py class">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">citegeist_model.</span></span><span class="sig-name descname"><span class="pre">CitegeistModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simulation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_expression_adata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">antibody_capture_adata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.__repr__" title="Permalink to this definition">#</a></dt>
<dd><p>Developer-friendly representation of the CitegeistModel.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.__str__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.__str__" title="Permalink to this definition">#</a></dt>
<dd><p>User-friendly representation of the CitegeistModel.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.append_gex_to_adata">
<span class="sig-name descname"><span class="pre">append_gex_to_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parquet_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pass_number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.append_gex_to_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.append_gex_to_adata" title="Permalink to this definition">#</a></dt>
<dd><p>Append gene expression layers from a Parquet file back into the gene_expression_adata object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.append_proportions_to_adata">
<span class="sig-name descname"><span class="pre">append_proportions_to_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">proportions_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'finetuned'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.append_proportions_to_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.append_proportions_to_adata" title="Permalink to this definition">#</a></dt>
<dd><p>Append cell type proportions to AnnData object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.cleanup">
<span class="sig-name descname"><span class="pre">cleanup</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.cleanup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.cleanup" title="Permalink to this definition">#</a></dt>
<dd><p>Free memory and clean up temporary data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.compute_expression_prior">
<span class="sig-name descname"><span class="pre">compute_expression_prior</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spotwise_profiles_pass1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_numbers_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_prior</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_expression_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.compute_expression_prior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.compute_expression_prior" title="Permalink to this definition">#</a></dt>
<dd><p>Compute global prior from pass 1 results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spotwise_profiles_pass1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – Dictionary mapping spot indices to profile matrices</p></li>
<li><p><strong>cell_type_numbers_array</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>) – Array of cell type proportions (N_spots × T_celltypes)</p></li>
<li><p><strong>lambda_prior</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Strength of prior influence (default: 1.0)</p></li>
<li><p><strong>min_expression_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Minimum expression to consider “active” (default: 0.1)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>{</dt><dd><p>‘global_prior’: np.ndarray,  # shape (T_celltypes, M_genes)
‘confidence_scores’: np.ndarray,  # shape (T_celltypes, M_genes)
‘expression_patterns’: Dict containing detailed statistics</p>
</dd>
</dl>
<p>}</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict[str, Any]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.copy_gex_to_protein_adata">
<span class="sig-name descname"><span class="pre">copy_gex_to_protein_adata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.copy_gex_to_protein_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.copy_gex_to_protein_adata" title="Permalink to this definition">#</a></dt>
<dd><p>Copy the number of spots in the gene expression AnnData object to the antibody capture AnnData object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.filter_gex">
<span class="sig-name descname"><span class="pre">filter_gex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nonzero_percentage</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_expression_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.filter_gex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.filter_gex" title="Permalink to this definition">#</a></dt>
<dd><p>Filter genes in the gene expression AnnData object based on user-defined criteria.</p>
<p>Filters genes that have:
1. A count &gt; 0 in at least <cite>nonzero_percentage</cite> of spots
2. Mean expression &gt; <cite>mean_expression_threshold</cite> in nonzero spots</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nonzero_percentage</strong> (<em>float</em>) – Minimum percentage of spots where a gene must have a count &gt; 0 (default: 1%)</p></li>
<li><p><strong>mean_expression_threshold</strong> (<em>float</em>) – Minimum mean expression value in nonzero spots</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.get_adata">
<span class="sig-name descname"><span class="pre">get_adata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.get_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.get_adata" title="Permalink to this definition">#</a></dt>
<dd><p>Retrieve the internal AnnData object for downstream analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The internal <cite>adata</cite> object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.global_clr">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">global_clr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epsilon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-06</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.global_clr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.global_clr" title="Permalink to this definition">#</a></dt>
<dd><p>Apply margin=2 CLR normalization (global geometric mean per marker).
:type matrix: 
:param matrix: Input matrix (spots x markers).
:type matrix: numpy.ndarray
:type epsilon: 
:param epsilon: Small constant to avoid division by zero.
:type epsilon: float</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>CLR-normalized matrix.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.load_cell_profile_dict">
<span class="sig-name descname"><span class="pre">load_cell_profile_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_profile_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.load_cell_profile_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.load_cell_profile_dict" title="Permalink to this definition">#</a></dt>
<dd><p>Load and validate the cell profile dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cell_profile_dict</strong> (<em>dict</em>) – Dictionary of cell type profiles.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.preprocess_antibody">
<span class="sig-name descname"><span class="pre">preprocess_antibody</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.preprocess_antibody"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.preprocess_antibody" title="Permalink to this definition">#</a></dt>
<dd><p>Preprocess antibody capture data:
- Winsorize extreme values.
- Apply Gaussian smoothing for local background correction.
- Apply global CLR normalization.
- Raise an error if NaNs or Infs are detected in the processed data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.preprocess_gex">
<span class="sig-name descname"><span class="pre">preprocess_gex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.preprocess_gex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.preprocess_gex" title="Permalink to this definition">#</a></dt>
<dd><p>Preprocess gene expression data with count-preserving normalization.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.register_gurobi">
<span class="sig-name descname"><span class="pre">register_gurobi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">license_file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.register_gurobi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.register_gurobi" title="Permalink to this definition">#</a></dt>
<dd><p>Configure Gurobi by setting only the license file path.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>license_file_path</strong> (<em>str</em>) – Path to the Gurobi license file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.row_normalize">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">row_normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.row_normalize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.row_normalize" title="Permalink to this definition">#</a></dt>
<dd><p>Row normalize a 2D NumPy array to a fixed target sum.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.run_cell_expression_pass1">
<span class="sig-name descname"><span class="pre">run_cell_expression_pass1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">radius</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_reg_gex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_enrichment_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_enrichment_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">checkpoint_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'checkpoints'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rerun</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.run_cell_expression_pass1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.run_cell_expression_pass1" title="Permalink to this definition">#</a></dt>
<dd><p>Run first pass of gene expression deconvolution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>radius</strong> (<em>float</em>) – Radius for neighbor detection</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Weight for spatial regularization</p></li>
<li><p><strong>lambda_reg_gex</strong> (<em>float</em>) – Weight for gene expression regularization</p></li>
<li><p><strong>global_enrichment_weight</strong> (<em>float</em>) – Weight for global expression enrichment (0-1)</p></li>
<li><p><strong>local_enrichment_weight</strong> (<em>float</em>) – Weight for local expression enrichment (0-1)</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of parallel workers</p></li>
<li><p><strong>checkpoint_interval</strong> (<em>int</em>) – Number of spots between checkpoints</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Directory for checkpoints</p></li>
<li><p><strong>rerun</strong> (<em>bool</em>) – Whether to rerun if results exist</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>{</dt><dd><p>‘spotwise_profiles’: Dict[int, np.ndarray],
‘dimensions’: Tuple[int, int, int]</p>
</dd>
</dl>
<p>}</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict[str, Any]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.run_cell_proportion_model">
<span class="sig-name descname"><span class="pre">run_cell_proportion_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_reg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_y_change</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">checkpoint_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.run_cell_proportion_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.run_cell_proportion_model" title="Permalink to this definition">#</a></dt>
<dd><p>Orchestrates the cell proportion optimization workflow.
Delegates optimization to <cite>optimize_cell_proportions</cite> and <cite>finetune_cell_proportions</cite> in <cite>gurobi_impl.py</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tolerance</strong> (<em>float</em>) – Convergence tolerance for EM algorithm</p></li>
<li><p><strong>max_iterations</strong> (<em>int</em>) – Maximum number of iterations</p></li>
<li><p><strong>lambda_reg</strong> (<em>float</em>) – Regularization strength</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – L1-L2 tradeoff factor (0 = L2, 1 = L1)</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of parallel workers for finetuning</p></li>
<li><p><strong>checkpoint_interval</strong> (<em>int</em>) – Number of spots between checkpoints during finetuning</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.split_adata">
<span class="sig-name descname"><span class="pre">split_adata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.split_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.split_adata" title="Permalink to this definition">#</a></dt>
<dd><p>Split the AnnData object into separate gene expression and antibody capture sub-objects
based on ‘feature_types’ in <cite>adata.var</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.validate_neighborhood_size">
<span class="sig-name descname"><span class="pre">validate_neighborhood_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">radius</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.validate_neighborhood_size"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.validate_neighborhood_size" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="citegeist_model.CitegeistModel.winsorize">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">winsorize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_percentile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_percentile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">95</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/citegeist_model.html#CitegeistModel.winsorize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#citegeist_model.CitegeistModel.winsorize" title="Permalink to this definition">#</a></dt>
<dd><p>Winsorize a 2D NumPy array.</p>
</dd></dl>

</dd></dl>

</section>
<section id="gurobi-implementation">
<h2>Gurobi Implementation<a class="headerlink" href="#gurobi-implementation" title="Permalink to this heading">#</a></h2>
<p>Core optimization functions using Gurobi solver.</p>
<span class="target" id="module-gurobi_impl"></span><dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.compute_global_prior">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">compute_global_prior</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spotwise_gene_expression_profiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_numbers_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_prior</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_expression_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#compute_global_prior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.compute_global_prior" title="Permalink to this definition">#</a></dt>
<dd><p>Compute global prior from pass 1 results using normalized expression patterns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spotwise_gene_expression_profiles</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]) – Dictionary mapping spot indices to profile matrices</p></li>
<li><p><strong>cell_type_numbers_array</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>) – Array of cell type proportions (N_spots × T_celltypes)</p></li>
<li><p><strong>lambda_prior</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Strength of prior (default: 1.0)</p></li>
<li><p><strong>min_expression_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Minimum expression to consider “active” (default: 0.1)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>global_prior: Prior matrix (T_celltypes × M_genes)</p></li>
<li><p>confidence_scores: Confidence in each prior value</p></li>
<li><p>expression_patterns: Summary of expression patterns</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict containing</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.deconvolute_local_cell_proportions">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">deconvolute_local_cell_proportions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">profile_based_antibody_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_reg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beta_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beta_vary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_beta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_y_change</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#deconvolute_local_cell_proportions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.deconvolute_local_cell_proportions" title="Permalink to this definition">#</a></dt>
<dd><p>Refine cell proportions for a single spot via local neighborhood optimization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spot_idx</strong> (<em>int</em>) – Index of the spot to refine in the AnnData object.</p></li>
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData containing spot-level spatial coordinates in obsm[‘spatial’].</p></li>
<li><p><strong>profile_based_antibody_data</strong> (<em>np.ndarray</em>) – (N x T) global antibody intensities for N spots, T cell types.</p></li>
<li><p><strong>radius</strong> (<em>float</em>) – Neighborhood radius for identifying neighbors.</p></li>
<li><p><strong>tolerance</strong> (<em>float</em>) – Convergence threshold for Y- and beta-updates (if beta_vary=True).</p></li>
<li><p><strong>lambda_reg</strong> (<em>float</em>) – Strength of elastic net regularization.</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – L1-L2 tradeoff for the elastic net (0 = L2, 1 = L1).</p></li>
<li><p><strong>beta_values</strong> (<em>Optional</em><em>[</em><em>np.ndarray</em><em>]</em>) – Global or initial local beta values (length T). If None and beta_vary=True,
local betas initialize at 1.0 each.</p></li>
<li><p><strong>beta_vary</strong> (<em>bool</em>) – If True, local betas are iteratively updated.
If False, beta_values remain fixed throughout optimization.</p></li>
<li><p><strong>normalize_beta</strong> (<em>bool</em>) – Whether to normalize beta values after updates.</p></li>
<li><p><strong>max_iterations</strong> (<em>int</em>) – Maximum iterations allowed for EM-like steps within this local function.</p></li>
<li><p><strong>max_y_change</strong> (<em>float</em>) – Maximum allowed change in Y values between iterations (default: 0.2).
Values are constrained to vary by at most this amount while staying in [0,1].</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Refined proportions (T,) for the specified spot, or None on failure.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.deconvolute_spot_with_neighbors_with_prior">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">deconvolute_spot_with_neighbors_with_prior</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_numbers_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_prior</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_prior_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_enrichment_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_enrichment_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#deconvolute_spot_with_neighbors_with_prior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.deconvolute_spot_with_neighbors_with_prior" title="Permalink to this definition">#</a></dt>
<dd><p>Deconvolute a spot with its neighbors, using both enrichment weights and optional prior.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.finetune_cell_proportions">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">finetune_cell_proportions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">profile_based_antibody_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_Y_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_beta_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_reg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_y_change</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beta_vary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">checkpoint_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'checkpoints'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rerun</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#finetune_cell_proportions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.finetune_cell_proportions" title="Permalink to this definition">#</a></dt>
<dd><p>Refine cell proportions using local neighborhood optimization with parallelization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>profile_based_antibody_data</strong> (<em>np.ndarray</em>) – (N x T) array of mapped antibody intensities.</p></li>
<li><p><strong>cell_type_names</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Ordered list of length T specifying cell type names.</p></li>
<li><p><strong>initial_Y_values</strong> (<em>np.ndarray</em>) – Initial cell proportion matrix of shape (N, T).</p></li>
<li><p><strong>initial_beta_values</strong> (<em>np.ndarray</em>) – Initial beta estimates (shape T,). This is passed for consistency,
but local solver decides how to use or ignore it based on beta_vary.</p></li>
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – AnnData object with spot-level spatial coordinates in obsm[‘spatial’].</p></li>
<li><p><strong>radius</strong> (<em>float</em>) – Radius for neighborhood-based local refinement.</p></li>
<li><p><strong>tolerance</strong> (<em>float</em>) – Convergence tolerance for the local optimization loops.</p></li>
<li><p><strong>lambda_reg</strong> (<em>float</em>) – Elastic net regularization strength for local solver.</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – L1-L2 tradeoff (0 = purely L2, 1 = purely L1) in local solver.</p></li>
<li><p><strong>max_iterations</strong> (<em>int</em>) – Maximum number of iterations for local solver.</p></li>
<li><p><strong>max_y_change</strong> (<em>float</em>) – Maximum allowed change in Y values between iterations (default: 0.2).
Values are constrained to vary by at most this amount while staying in [0,1].</p></li>
<li><p><strong>beta_vary</strong> (<em>bool</em>) – If True, each spot’s local solver is allowed to update betas;
if False, betas remain fixed at the values passed in initial_beta_values.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of parallel workers. If None, uses os.cpu_count().</p></li>
<li><p><strong>checkpoint_interval</strong> (<em>int</em>) – Number of spots between checkpoints.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Directory for checkpoints.</p></li>
<li><p><strong>rerun</strong> (<em>bool</em>) – Whether to rerun if results exist.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A new (N x T) array of refined Y-values, obtained from a single pass
of local refinements.</p></li>
<li><p>The original beta_values array, returned for interface consistency.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[np.ndarray, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.log_marker_gene_patterns">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">log_marker_gene_patterns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">zero_patterns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_genes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#log_marker_gene_patterns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.log_marker_gene_patterns" title="Permalink to this definition">#</a></dt>
<dd><p>Log detailed patterns for marker genes.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.map_antibodies_to_profiles">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">map_antibodies_to_profiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_profile_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#map_antibodies_to_profiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.map_antibodies_to_profiles" title="Permalink to this definition">#</a></dt>
<dd><p>Map antibody capture data to predefined cell type profiles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Antibody capture AnnData object.</p></li>
<li><p><strong>cell_profile_dict</strong> (<em>dict</em>) – Dictionary mapping cell types to antibody markers.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Profile-based antibody data matrix (N_spots x T_cell_types).
list: List of cell type names (to ensure column order).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.normalize_counts">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">normalize_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_highly_expressed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#normalize_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.normalize_counts" title="Permalink to this definition">#</a></dt>
<dd><p>Normalize counts while preserving integer values and relative proportions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object</p></li>
<li><p><strong>target_sum</strong> – Target sum for each cell/spot</p></li>
<li><p><strong>exclude_highly_expressed</strong> – Whether to exclude highly expressed genes</p></li>
<li><p><strong>max_fraction</strong> – Maximum fraction for highly expressed genes</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.optimize_cell_proportions">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">optimize_cell_proportions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">profile_based_antibody_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_reg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_beta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#optimize_cell_proportions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.optimize_cell_proportions" title="Permalink to this definition">#</a></dt>
<dd><p>Perform EM-based optimization for cell type proportions using Gurobi.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>profile_based_antibody_data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>) – N x T matrix of mapped antibody data</p></li>
<li><p><strong>cell_type_names</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of cell type names</p></li>
<li><p><strong>tolerance</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Convergence tolerance for EM algorithm</p></li>
<li><p><strong>max_iterations</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of iterations</p></li>
<li><p><strong>lambda_reg</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Regularization strength</p></li>
<li><p><strong>alpha</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – L1-L2 tradeoff factor (0 = L2, 1 = L1)</p></li>
<li><p><strong>normalize_beta</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to normalize beta values</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Y_values (N x T), beta_values (T,)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[np.ndarray, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.optimize_gene_expression">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">optimize_gene_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deconvolution_expression_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_numbers_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filtered_adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_enrichment_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_enrichment_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_prior</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_prior_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">checkpoint_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'checkpoints'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rerun</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#optimize_gene_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.optimize_gene_expression" title="Permalink to this definition">#</a></dt>
<dd><p>Optimize gene expression with enrichment weights and prior guidance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_name</strong> (<em>str</em>) – Name of the sample</p></li>
<li><p><strong>deconvolution_expression_data</strong> (<em>np.ndarray</em>) – Gene expression data (N_spots x M_genes)</p></li>
<li><p><strong>cell_type_numbers_array</strong> (<em>np.ndarray</em>) – Cell type proportions (N_spots x T_celltypes)</p></li>
<li><p><strong>filtered_adata</strong> (<em>sc.AnnData</em>) – Filtered AnnData object containing gene expression data</p></li>
<li><p><strong>radius</strong> (<em>float</em>) – Radius for neighbor detection</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Weight for spatial regularization</p></li>
<li><p><strong>lambda_reg_gex</strong> (<em>float</em>) – Weight for gene expression regularization</p></li>
<li><p><strong>global_enrichment_weight</strong> (<em>float</em>) – Weight for global expression enrichment (0-1)</p></li>
<li><p><strong>local_enrichment_weight</strong> (<em>float</em>) – Weight for local expression enrichment (0-1)</p></li>
<li><p><strong>global_prior</strong> (<em>np.ndarray</em><em>, </em><em>optional</em>) – Global prior matrix for guidance</p></li>
<li><p><strong>lambda_prior_weight</strong> (<em>float</em>) – Weight for prior guidance</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of parallel workers</p></li>
<li><p><strong>checkpoint_interval</strong> (<em>int</em>) – Number of spots between checkpoints</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Directory for checkpoints</p></li>
<li><p><strong>rerun</strong> (<em>bool</em>) – Whether to rerun if results exist</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>{</dt><dd><p>‘spotwise_profiles’: Dict[int, np.ndarray],
‘dimensions’: Tuple[int, int, int]</p>
</dd>
</dl>
<p>}</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict[str, Any]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.scale_genes">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">scale_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">expression_matrix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#scale_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.scale_genes" title="Permalink to this definition">#</a></dt>
<dd><p>Scale each gene independently to [0,1] range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>expression_matrix</strong> (<em>np.ndarray</em>) – Spots x Genes matrix</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(scaled_matrix, gene_mins, gene_maxs)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.unscale_genes">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">unscale_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scaled_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_mins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_maxs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#unscale_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.unscale_genes" title="Permalink to this definition">#</a></dt>
<dd><p>Reverse the gene-wise scaling transformation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scaled_matrix</strong> (<em>np.ndarray</em>) – Scaled matrix</p></li>
<li><p><strong>gene_mins</strong> (<em>np.ndarray</em>) – Original minimum values per gene</p></li>
<li><p><strong>gene_maxs</strong> (<em>np.ndarray</em>) – Original maximum values per gene</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Unscaled matrix</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gurobi_impl.validate_prior_effect">
<span class="sig-prename descclassname"><span class="pre">gurobi_impl.</span></span><span class="sig-name descname"><span class="pre">validate_prior_effect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spotwise_profiles_pass1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spotwise_profiles_pass2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_prior</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/gurobi_impl.html#validate_prior_effect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gurobi_impl.validate_prior_effect" title="Permalink to this definition">#</a></dt>
<dd><p>Compare pass1 and pass2 results to verify prior influence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spotwise_profiles_pass1</strong> (<em>dict</em>) – First pass results {spot_idx: profile_matrix}</p></li>
<li><p><strong>spotwise_profiles_pass2</strong> (<em>dict</em>) – Second pass results {spot_idx: profile_matrix}</p></li>
<li><p><strong>global_prior</strong> (<em>np.ndarray</em>) – Global prior matrix (T x M)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing validation metrics</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Permalink to this heading">#</a></h2>
<p>Helper functions and utilities for CITEgeist.</p>
<span class="target" id="module-utils"></span><dl class="py function">
<dt class="sig sig-object py" id="utils.assert_neighborhood_size">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">assert_neighborhood_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_profile_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_spots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#assert_neighborhood_size"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.assert_neighborhood_size" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.benchmark_cell_proportions">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">benchmark_cell_proportions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">true_proportions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">predicted_proportions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#benchmark_cell_proportions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.benchmark_cell_proportions" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate performance metrics for cell type proportion predictions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>true_proportions</strong> (<em>np.ndarray</em>) – Ground truth cell type proportions matrix</p></li>
<li><p><strong>predicted_proportions</strong> (<em>np.ndarray</em>) – Predicted cell type proportions matrix</p></li>
<li><p><strong>cell_type_names</strong> (<em>list</em>) – Names of cell types corresponding to matrix columns</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing various performance metrics</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.calculate_expression_metrics">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">calculate_expression_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ground_truth_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">predictions_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'range'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pass_number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#calculate_expression_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.calculate_expression_metrics" title="Permalink to this definition">#</a></dt>
<dd><p>Calculate performance metrics for gene expression predictions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ground_truth_dir</strong> (<em>str</em>) – Directory containing ground truth CSV files</p></li>
<li><p><strong>predictions_dir</strong> (<em>str</em>) – Directory containing prediction CSV files</p></li>
<li><p><strong>normalize</strong> (<em>str</em>) – Normalization method for NRMSE (‘range’ or ‘mean’)</p></li>
<li><p><strong>pass_number</strong> (<em>int</em><em>, </em><em>optional</em>) – If specified, look for predictions in pass-specific subdirectory</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing performance metrics per cell type and overall statistics</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.cleanup_memory">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">cleanup_memory</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#cleanup_memory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.cleanup_memory" title="Permalink to this definition">#</a></dt>
<dd><p>Force garbage collection to free memory.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.export_anndata_layers">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">export_anndata_layers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pass_number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#export_anndata_layers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.export_anndata_layers" title="Permalink to this definition">#</a></dt>
<dd><p>Export all layers of an AnnData object to separate CSV files.
Creates separate folders for different passes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object containing the layers to export</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Base directory where CSV files will be saved</p></li>
<li><p><strong>pass_number</strong> (<em>int</em><em>, </em><em>optional</em>) – If specified, only export layers from this pass</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.find_fixed_radius_neighbors">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">find_fixed_radius_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#find_fixed_radius_neighbors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.find_fixed_radius_neighbors" title="Permalink to this definition">#</a></dt>
<dd><p>Find neighbors within a fixed radius of a given spot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spot_index</strong> (<em>int</em>) – Index of the central spot in the AnnData object.</p></li>
<li><p><strong>adata</strong> (<em>AnnData</em>) – Spatial transcriptomics dataset with obsm[‘spatial’] containing spot coordinates.</p></li>
<li><p><strong>radius</strong> (<em>float</em>) – Fixed radius to identify neighboring spots.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(central_spot_name, list of neighbor spot names)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.get_neighbors_with_fixed_radius">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">get_neighbors_with_fixed_radius</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#get_neighbors_with_fixed_radius"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.get_neighbors_with_fixed_radius" title="Permalink to this definition">#</a></dt>
<dd><p>Get indices of neighboring spots based on a fixed radius around the central spot.</p>
<p>Parameters:
- spot_index (int): The index of the central spot.
- adata (AnnData): The AnnData object with spatial coordinates in obsm[‘spatial’].
- radius (float): Fixed radius for finding neighbors.
- include_center (bool): Whether to include the central spot in the neighbor list.</p>
<p>Returns:
- List of indices representing the central spot and its neighbors.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.plot_neighbors_with_fixed_radius">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">plot_neighbors_with_fixed_radius</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_spots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#plot_neighbors_with_fixed_radius"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.plot_neighbors_with_fixed_radius" title="Permalink to this definition">#</a></dt>
<dd><p>Plot neighbors for multiple random central spots using <cite>sc.pl.spatial</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Spatial transcriptomics dataset with obsm[‘spatial’].</p></li>
<li><p><strong>radius</strong> (<em>float</em>) – Fixed radius to identify neighboring spots.</p></li>
<li><p><strong>num_spots</strong> (<em>int</em>) – Number of random spots to visualize.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Displays a series of spatial plots showing neighbors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.save_results_to_output">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">save_results_to_output</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">results</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#save_results_to_output"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.save_results_to_output" title="Permalink to this definition">#</a></dt>
<dd><p>Save results as a CSV file.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.setup_logging">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">setup_logging</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#setup_logging"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.setup_logging" title="Permalink to this definition">#</a></dt>
<dd><p>Set up dynamic logging.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="utils.validate_cell_profile_dict">
<span class="sig-prename descclassname"><span class="pre">utils.</span></span><span class="sig-name descname"><span class="pre">validate_cell_profile_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_profile_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/utils.html#validate_cell_profile_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#utils.validate_cell_profile_dict" title="Permalink to this definition">#</a></dt>
<dd><p>Validate the structure of a cell profile dictionary.</p>
</dd></dl>

</section>
</section>


                </article>
              
              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="notebooks/vignette_4_external_tools.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Vignette 4: Integration with External Tools</p>
      </div>
    </a>
    <a class="right-next"
       href="contributing.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Contributing</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist-model">CITEgeist Model</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel"><code class="docutils literal notranslate"><span class="pre">CitegeistModel</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.__repr__"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.__repr__()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.__str__"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.__str__()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.append_gex_to_adata"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.append_gex_to_adata()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.append_proportions_to_adata"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.append_proportions_to_adata()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.cleanup"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.cleanup()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.compute_expression_prior"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.compute_expression_prior()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.copy_gex_to_protein_adata"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.copy_gex_to_protein_adata()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.filter_gex"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.filter_gex()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.get_adata"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.get_adata()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.global_clr"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.global_clr()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.load_cell_profile_dict"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.load_cell_profile_dict()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.preprocess_antibody"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.preprocess_antibody()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.preprocess_gex"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.preprocess_gex()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.register_gurobi"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.register_gurobi()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.row_normalize"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.row_normalize()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.run_cell_expression_pass1"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.run_cell_expression_pass1()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.run_cell_proportion_model"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.run_cell_proportion_model()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.split_adata"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.split_adata()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.validate_neighborhood_size"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.validate_neighborhood_size()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#citegeist_model.CitegeistModel.winsorize"><code class="docutils literal notranslate"><span class="pre">CitegeistModel.winsorize()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi-implementation">Gurobi Implementation</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.compute_global_prior"><code class="docutils literal notranslate"><span class="pre">compute_global_prior()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.deconvolute_local_cell_proportions"><code class="docutils literal notranslate"><span class="pre">deconvolute_local_cell_proportions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.deconvolute_spot_with_neighbors_with_prior"><code class="docutils literal notranslate"><span class="pre">deconvolute_spot_with_neighbors_with_prior()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.finetune_cell_proportions"><code class="docutils literal notranslate"><span class="pre">finetune_cell_proportions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.log_marker_gene_patterns"><code class="docutils literal notranslate"><span class="pre">log_marker_gene_patterns()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.map_antibodies_to_profiles"><code class="docutils literal notranslate"><span class="pre">map_antibodies_to_profiles()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.normalize_counts"><code class="docutils literal notranslate"><span class="pre">normalize_counts()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.optimize_cell_proportions"><code class="docutils literal notranslate"><span class="pre">optimize_cell_proportions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.optimize_gene_expression"><code class="docutils literal notranslate"><span class="pre">optimize_gene_expression()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.scale_genes"><code class="docutils literal notranslate"><span class="pre">scale_genes()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.unscale_genes"><code class="docutils literal notranslate"><span class="pre">unscale_genes()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gurobi_impl.validate_prior_effect"><code class="docutils literal notranslate"><span class="pre">validate_prior_effect()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#utilities">Utilities</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.assert_neighborhood_size"><code class="docutils literal notranslate"><span class="pre">assert_neighborhood_size()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.benchmark_cell_proportions"><code class="docutils literal notranslate"><span class="pre">benchmark_cell_proportions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.calculate_expression_metrics"><code class="docutils literal notranslate"><span class="pre">calculate_expression_metrics()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.cleanup_memory"><code class="docutils literal notranslate"><span class="pre">cleanup_memory()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.export_anndata_layers"><code class="docutils literal notranslate"><span class="pre">export_anndata_layers()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.find_fixed_radius_neighbors"><code class="docutils literal notranslate"><span class="pre">find_fixed_radius_neighbors()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.get_neighbors_with_fixed_radius"><code class="docutils literal notranslate"><span class="pre">get_neighbors_with_fixed_radius()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.plot_neighbors_with_fixed_radius"><code class="docutils literal notranslate"><span class="pre">plot_neighbors_with_fixed_radius()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.save_results_to_output"><code class="docutils literal notranslate"><span class="pre">save_results_to_output()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.setup_logging"><code class="docutils literal notranslate"><span class="pre">setup_logging()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#utils.validate_cell_profile_dict"><code class="docutils literal notranslate"><span class="pre">validate_cell_profile_dict()</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div class="tocsection sourcelink">
    <a href="_sources/api.rst.txt">
      <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
  </div>
</div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">
  <p class="copyright">
    
      © Copyright 2025, Lee/Oesterreich Lab.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">
  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 5.3.0.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item"><p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.13.3.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>